src/tls_stream.rs

```diff
@@ -200,6 +201,35 @@ pub enum HandshakeError<S> {
     Interrupted(MidHandshakeTlsStream<S>),
 }
 
+/// A struct used to wrap various cert chain validation results for callback processing. 
+pub struct CertValidationResult {
+    chain :CertChainContext,
+    res :io::Result<()>,
+    chain_index :i32,
+    element_index :i32,
+    extra_policy_status : (winapi::LPCSTR, *const winapi::c_void),
+}
+
+impl CertValidationResult {
+
+    /// Returns the certificate that failed validation if applicable
+    pub fn failed_certificate(&self) -> Option<CertContext> {
+        if let Some(cert_chain) = self.chain.get_chain(self.chain_index as usize) {
+            return cert_chain.get(self.element_index as usize);
+        }
+        None
+    }
+    
+    // Returns the final certificate chain in the certificate context if applicable
+    pub fn chain(&self) -> Option<CertChain> {
+        self.chain.final_chain()
+    }
+    
+    pub fn result(&self) -> &io::Result<()> {
```

Comment:
`io::Error` isn't `Clone`, so we'll want this to return the `Result` by value so that it can then be returned from the closure - we should probably store the error code directly and construct the `Error` on each call.